mdsuite/calculators/radial_distribution_function.py

```diff
@@ -575,14 +540,27 @@ def compute_species_values(indices, atoms_per_batch, start, d_ij):
             else:
                 particles_list = [len(self.experiment.species[item]['indices']) for item in self.experiment.species]
 
-            for species_indices, key in self.mini_get_pair_indices(indices, n_atoms,
-                                                                   atoms_per_batch, start,
-                                                                   self.index_list, particles_list):
-                # iterate over the permutations between the species
-                distance_between_species = tf.gather(d_ij, species_indices)
-                distance_between_species = self._apply_system_cutoff(distance_between_species, self.cutoff)
-                bin_data = tf.histogram_fixed_width(distance_between_species, self.bin_range, self.number_of_bins)
-                rdf[key] = bin_data
+            indices = tf.transpose(indices)
+
+            for tuples in itertools.combinations_with_replacement(self.index_list, 2):
+                start_ = (sum(particles_list[:tuples[0]]) - start_batch, sum(particles_list[:tuples[1]]))
+                size_ = (
+                    particles_list[tuples[0]],
+                    particles_list[tuples[1]]
+                )
+
+                names = self._get_species_names(tuples)
+
+                # select the indices that are within the boundaries of the current species / molecule
+                mask_1 = tf.math.logical_and(indices[:, 0] > start_[0], indices[:, 0] < start_[0] + size_[0])
```

Comment:
This is fast, but is there a faster way without the logic? I am not sure how tf handles logic but it is typically slower than building a binary mask.

Reply:
I don't know.
```py
mask_1 = (indices[:, 0] > start[0]) & (indices[:, 0] < stop[0])
mask_2 = (indices[:, 1] > start[1]) & (indices[:, 1] < stop[1])
```

This is just another notation that yields the same speeds I tried